Add documentation for CaseDescription (#130)
ESSS/alfasim-sdk: docs/source/alfacase/case_description.example.rst

    from alfasim_sdk import (
        CaseDescription,
        NodeDescription,
        MassSourceNodePropertiesDescription,
        PressureNodePropertiesDescription,
    )

    CaseDescription(
        # Omitting fields that configure Options and Pvt
        nodes=[
            NodeDescription(
                name="Inlet",
                node_type=NodeCellType.MassSource,
                mass_source_properties=MassSourceNodePropertiesDescription(
                    mass_flow_rates={
                        "gas": Scalar(0.0, "kg/s"),
                        "oil": Scalar(0.0, "kg/s"),
                    },
                ),
            ),
            NodeDescription(
                name="Outlet",
                node_type=NodeCellType.Pressure,
                pressure_properties=PressureNodePropertiesDescription(
                    pressure=Scalar(50.0, "bar"),
                    volume_fractions={
                        "gas": Scalar(1.0, "-"),
                        "oil": Scalar(1.0, "-"),
                    },
                ),
            ),
        ],
    )

Pipes
-----

Let's add a Pipe to the application through the ``pipes`` fields.

> The ``pipes`` accept a list of :class:`PipeDescription <alfasim_sdk.PipeDescription>` definitions that connects two nodes.

Reviewer: ```suggestion
The ``pipes`` attribute accept a list of :class:`PipeDescription <alfasim_sdk.PipeDescription>` definitions that connects two nodes.
```

Example
=======

In this section, we will walk trough the creation of a simple network, with two nodes and one pipe, to illustrate how to
create a project and how to interpret the definitions available on :ref:`alfacase-reference-section`.

The root level of a ``Description`` is always the class :class:`CaseDescription <alfasim_sdk.CaseDescription>`
which has the following attributes:

.. code-block:: python

    class CaseDescription
        name: Optional[str] = None
        physics: PhysicsDescription = PhysicsDescription()
        time_options: TimeOptionsDescription = TimeOptionsDescription()
        numerical_options: NumericalOptionsDescription = NumericalOptionsDescription()
        ipr_models: IPRModelsDescription = IPRModelsDescription()
        pvt_models: PvtModelsDescription = PvtModelsDescription()
        tracers: TracersDescription = TracersDescription()
        outputs: CaseOutputDescription = CaseOutputDescription()
        pipes: List[PipeDescription] = []
        nodes: List[NodeDescription] = []
        wells: List[WellDescription] = []
        materials: List[MaterialDescription] = []
        walls: List[WallDescription] = []

> In the following section, this steps will be covered:

Reviewer: ```suggestion
In the following section, these steps will be covered:
```

- How to configure the project options.
- How to add a PVtModel.
- How to add Nodes.
- How to add Pipes.
- How to add output curves.

.. note::

    For the full reference check the :ref:`alfacase-reference-section`

Project Options
---------------

We will start the example by configuring the project options, as discussed in the previous section,
the root level of an ``ALFacase`` file is the :class:`CaseDescription <alfasim_sdk.CaseDescription>` and it has the following options:

-  ``physics``: PhysicsDescription = PhysicsDescription()
-  ``time_options``: TimeOptionsDescription = TimeOptionsDescription()
-  ``numerical_options``: NumericalOptionsDescription = NumericalOptionsDescription()

.. note::

>     The name before the colon is the attribute name, the value after the colon is the type and the value after the

Reviewer 2: I like the idea of a disclaimer at the start. It don't need to be big or any thing.

>     The name before the colon is the attribute name, the value after the colon is the type and the value after the

Reviewer 3: I propose:

```
rst

.. note::

   This guide assumes the reader is familiar with `Python 3 <https://www.python.org/>`__.
```

>     The name before the colon is the attribute name, the value after the colon is the type and the value after the

Author: n my "defense" I thought of the following scenario.
The user knows Python, but has never seen a type hint or used something like dataclass or attrs using typehints.

In this sense, my intention was more on the way to make the user aware of how to interpret the definition provided correctly.

>     The name before the colon is the attribute name, the value after the colon is the type and the value after the

Reviewer 3: I still strongly suggest to drop it.

>     The name before the colon is the attribute name, the value after the colon is the type and the value after the

Reviewer 2: ` This guide assumes the reader is familiar with `Python 3 <https://www.python.org/>`__ and `type hints`<link to type hint docs on PEP>__.`

            hydrodynamic_model=HydrodynamicModelType.FourFields,
        ),
        numerical_options=NumericalOptionsDescription(tolerance=1e-4),
        time_options=TimeOptionsDescription(
            minimum_timestep=Scalar(1e-4, "s"),
            maximum_timestep=Scalar(0.5, "s"),
            final_time=Scalar(1.0, "s"),
        ),
    )

PVT Model
---------

The second step will add a ``PVTModel`` to the project and configure it as the default PVT for the entire project.

The ``pvt_models`` field from :class:`CaseDescription <alfasim_sdk.CaseDescription>` needs to be configured with the
definition provided by :class:`PvtModelsDescription <alfasim_sdk.PvtModelsDescription>`.

The :class:`PvtModelsDescription <alfasim_sdk.PvtModelsDescription>` is the root configuration of all PVTs over the application,
its possible to add new PVTs and defined one of them to be used automatically on all fields that require a
PVT through the attribute ``default_model``

For this example, a PVT will be created from a ``.tab`` file and the path to the file is relative to the ``.alfacase`` file generated.
And for this, we need to populate the ``tables`` field with the PVT name and a file.

> .. admonition:: About the tab file path

Author: hmm... I have second thoughts about your statement 😋 

Are you think @oliveira-mauricio @prusse-martin

> .. admonition:: About the tab file path

Reviewer 2: I think any one with a minimal knowledge of programing know what a relative path is.

> .. admonition:: About the tab file path

Reviewer: Yeah I think just saying it is relative to the alfacase file is enough.

    Considering that a ``.alfacase`` file is located at ``C:\Users\alfasim`` and the ``table`` section is configured with
    a relative path to ``my_pvt_file.tab``.

    In this case, the application will look for the ``.tab`` file at  ``C:\Users\alfasim\my_pvt_file.tab``

    .. code-block:: yaml

        tables:
           # PVT name  : file path
          'Pvt1': my_pvt_file.tab

.. note::

    The PVT name must be unique.

    Check the :ref:`alfacase-reference-pvt-section` section for a detailed description of each PVT type option.

.. rubric:: Example

.. code-block:: python

    from alfasim_sdk import (
        CaseDescription,
        PvtModelsDescription,
    )

    CaseDescription(
        # Omitting fields that configure option
        pvt_models=PvtModelsDescription(
            default_model="Pvt1",
            tables={"Pvt1": "my_pvt_file.tab"},
        ),
    )

Nodes
-----

> The third step will add two different types of nodes, a mass source node, and a pressure node.

Reviewer: ```suggestion
The third step will add two different types of nodes: a mass source node and a pressure node.
```

> All nodes that will be used on the application need to be added over the ``nodes`` section of the

Reviewer 3: ```suggestion
Nodes are added over the ``nodes`` section of the
```

:class:`CaseDescription <alfasim_sdk.CaseDescription>`.

> The :class:`NodeDescription <alfasim_sdk.NodeDescription>` is responsible to configure several types of nodes

Reviewer 3: ```suggestion
The :class:`NodeDescription <alfasim_sdk.NodeDescription>` class configures several types of nodes
```

through the ``node_type`` field and their respective property fields.

> For example, when the ``node_type`` is  ``NodeCellType.MassSource``, besides the fields from :class:`NodeDescription <alfasim_sdk.NodeDescription>`

Reviewer 3: Let's change this two paragraphs in bullet points:

```rst
* When ``node_type`` is  ``NodeCellType.MassSource``, ``mass_source_properties`` will be taken in account.

* When ``node_type`` is ``NodeCellType.Pressure``, ``pressure_properties`` will be taken in account.

The other properties not related to the ``node_type`` are read by the application, but are not considered for the solution.
```

.. rubric:: Example

.. code-block:: python

    from alfasim_sdk import (
        CaseDescription,
        NodeDescription,
        MassSourceNodePropertiesDescription,
        PressureNodePropertiesDescription,
    )

    CaseDescription(
        # Omitting fields that configure Options and Pvt
        nodes=[
            NodeDescription(
                name="Inlet",
                node_type=NodeCellType.MassSource,
                mass_source_properties=MassSourceNodePropertiesDescription(
                    mass_flow_rates={
                        "gas": Scalar(0.0, "kg/s"),
                        "oil": Scalar(0.0, "kg/s"),
                    },
                ),
            ),
            NodeDescription(
                name="Outlet",
                node_type=NodeCellType.Pressure,
                pressure_properties=PressureNodePropertiesDescription(
                    pressure=Scalar(50.0, "bar"),
                    volume_fractions={
                        "gas": Scalar(1.0, "-"),
                        "oil": Scalar(1.0, "-"),
                    },
                ),
            ),
        ],
    )

Pipes
-----

> The fourth step will add a Pipe to the application through the ``pipes`` fields.

Reviewer: We could remove references to this "steps order" in case we add something in between later.

```suggestion
Let's add a Pipe to the application through the ``pipes`` fields.
```

> The ``pipes`` accept a list of :class:`PipeDescription <alfasim_sdk.PipeDescription>` definitions that connects two nodes.

Reviewer 3: ```suggestion
The ``pipes`` attribute accepts a list of :class:`PipeDescription <alfasim_sdk.PipeDescription>` definitions that connects two nodes.
```

> The connection occurs with the fields ``source`` and ``target`` and to configure these fields, it is only necessary

Reviewer 3: ```suggestion
The connection is defined by the ``source`` and ``target`` fields, and to configure them it is only necessary
```

.. note::

    Check the :ref:`alfacase-reference-pipe-section` section for a detailed description of the attributes available.

.. rubric:: Example

.. code-block:: python

    from alfasim_sdk import (
        CaseDescription,
        PipeDescription,
        ProfileDescription,
        LengthAndElevationDescription,
        PipeSegmentsDescription,
    )

    CaseDescription(
        # Omitting fields that configure Options, Pvt and Node
        pipes=[
            PipeDescription(
                name="pipe",
                source="Inlet",
                target="Outlet",
                profile=ProfileDescription(
                    length_and_elevation=LengthAndElevationDescription(
                        length=Array([0.0, 15.0, 30.0, 30.0, 15.0], "m"),
                        elevation=Array([0.0, 15.0, 30.0, 30.0, 15.0], "m"),
                    ),
                ),
                segments=PipeSegmentsDescription(
                    start_positions=Array([0.0], "m"),
                    diameters=Array([0.1], "m"),
                    roughnesses=Array([5e-05], "m"),
                ),
            ),
        ],
    )

Output
------

> The final step for our example will add a trend and a profile for our project.

Reviewer 3: ```suggestion
The final step for our example will configure a trend and a profile for our project.
```

As indicated on :class:`CaseDescription <alfasim_sdk.CaseDescription>`, the ``outputs`` field must be filled with

> the definition of :class:`CaseOutputDescription <alfasim_sdk.CaseOutputDescription>` which allows the configuration of ``trends``

Reviewer 3: ```suggestion
a :class:`CaseOutputDescription <alfasim_sdk.CaseOutputDescription>` instance which configures ``trends``
```

    shows all the available curves that can be used.

.. rubric:: Example

.. code-block:: python

    from alfasim_sdk import (
        CaseDescription,
        CaseOutputDescription,
        TrendOutputDescription,
        OutputAttachmentLocation,
        ProfileOutputDescription,
    )

    CaseDescription(
        # Omitting fields that configure Options, Pvt, Node and Pipes
        outputs=CaseOutputDescription(
            trends=[
                TrendOutputDescription(
                    element_name="pipe",
                    curve_names=["oil mass flow rate"],
                    position=Scalar(100.0, "m"),
                    location=OutputAttachmentLocation.Main,
                )
            ],
            trend_frequency=Scalar(0.1, "s"),
            profiles=[
                ProfileOutputDescription(
                    curve_names=["pressure"],
                    element_name="pipe",
                    location=OutputAttachmentLocation.Main,
                )
            ],
            profile_frequency=Scalar(0.1, "s"),
        ),
    )

Full Case
---------

> This section brings together all the previous sections, showing the full example that can be now used

Reviewer 3: This is a bit confusing, because just the classes are not enough to be imported by the application; I suggest to just merge this section and the last.

Author: Sure I will change it to:

```
This section brings together all the previous sections, showing the full example of a Description with a project configuration being converted to a ``.alfacase`` file, ready to be imported by the application.

Reviewer: Yes, good idea.
